### PR TITLE
activate foxpass sudoers

### PR DESCRIPTION
**Pull request by asakapab0i**
## Description of the change

> Description here

## Type of change
- [ ] Bug fix
- [x] New feature

### Testing

- [x] Testing information has been added - test cases checklist and steps followed for testing.
- [ ] Testing screenshot(s) attached for more information.

Tested in the following distros

- CentOS 7
- CentOS 8
- Ubuntu 20.04
- Ubuntu 18.04
- Amazon Linux 2

Debug option output, it will be displayed right before the service restart

```
--- Old /etc/sssd/sssd.conf
+++ New /etc/sssd/sssd.conf
@@ -8,7 +8,12 @@
 chpass_provider = ldap
 ldap_uri = ldaps://ldap.foxpass.com
 ldap_id_use_start_tls = False
-ldap_tls_cacertdir = /etc/openldap/cacerts
+ldap_tls_reqcert = demand
+ldap_default_bind_dn = cn=xxx,dc=xxx,dc=xxx
+ldap_tls_cacert = /etc/ssl/certs/ca-bundle.crt
+ldap_opt_timeout = 6
+enumerate = True
+ldap_default_authtok = xxx
 [sssd]
 services = nss, pam, autofs

--- Old /etc/ssh/sshd_config
+++ New /etc/ssh/sshd_config
@@ -137,3 +137,6 @@
 #      AllowTcpForwarding no
 #      PermitTTY no
 #      ForceCommand cvs server
+
+AuthorizedKeysCommand          /usr/local/sbin/foxpass_ssh_keys.sh
+AuthorizedKeysCommandUser      root
--- Old /etc/sudoers
+++ New /etc/sudoers
@@ -107,7 +107,7 @@
 %wheel ALL=(ALL)       ALL

 ## Same thing without a password
-# %wheel       ALL=(ALL)       NOPASSWD: ALL
+%wheel ALL=(ALL)       NOPASSWD:ALL

 ## Allows members of the users group to mount and unmount the
 ## cdrom as root
--- Old /etc/sudoers.d/95-foxpass-sudo
+++ New /etc/sudoers.d/95-foxpass-sudo
@@ -0,0 +1,2 @@
+# Adding Foxpass group to sudoers
```

### CODE REVIEW

File: linux/ubuntu/20.04/foxpass_setup.py [hunk 1]
```diff
@@ -239,6 +291,12 @@ def fix_eic():
         os.system('systemctl daemon-reload')
 
 
+def check_sudo_passwd():
+    result = os.popen('passwd --status root').read().split(' ')
+    if result and result[1] != 'P':
+        sys.exit('Please set the password of the `root` user before enabling ldap sudoers. E.g sudo passwd root')
```
Comment (aren) [inline linux/ubuntu/20.04/foxpass_setup.py:297]: why is this needed?

Comment (asakapab0i) [reply linux/ubuntu/20.04/foxpass_setup.py:297]: Hi @aren , we cannot install `sudo-ldap` package if the root user's password is not set. 

This only happens in Ubuntu.

File: linux/ubuntu/20.04/foxpass_setup.py [hunk 2]
```diff
@@ -213,6 +219,52 @@ def fix_nsswitch():
...
 
+def write_ldap_sudoers(uris, basedn, binddn, bindpw, sudoers_timed, bind_timelimit, query_timelimit):
+    check_sudo_passwd()
+    return_code = os.system('DEBIAN_FRONTEND=noninteractive apt-get install -y sudo-ldap')
+    if return_code != 0:
+        # bitshift right 8 to get rid of the signal portion of the return code
+        sys.exit(return_code >> 8)
+    with open('/etc/sudo-ldap.conf', "w") as w:
+        content = """\
+#
```
Comment (grk-it) [inline linux/ubuntu/20.04/foxpass_setup.py:230]: Suggestion: I think it would be useful to show the user the changes that were made to the configuration files. It could be an optional one and can be based on user inputting `yes` for a message `would you like to know the details of the modifications made by the script`

File: linux/centos/7/foxpass_setup.py [hunk 3]
```diff
@@ -26,6 +26,7 @@
```
Comment (grk-it) [inline linux/centos/7/foxpass_setup.py:26]: Thanks for doing this Bryan. This looks good. Diff gives a good idea to the user regarding the changes made by the script. In fact we could also have a `--dry-run` which provides the changes that will be done when executed in real. But this is good for now :)

File: docker/Docker_amazonlinux2 [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+ENV DEBIAN_FRONTEND=noninteractive
+
+RUN yum -y update && yum -y install epel-release python-pip openssh-server sudo
+#RUN yum --enablerepo=epel install -y python-pip
+RUN python -m pip install urllib3
+
+COPY linux/amzn/2.0/foxpass_setup.py /
+
+ENTRYPOINT python foxpass_setup.py --base-dn {BASE_DN} --bind-user {BIND_USER} --bind-pw {BIND_PW} --api-key {API_KEY} --enable-ldap-sudoers --debug
```
Comment (grk-it) [inline docker/Docker_amazonlinux2:16]: Could we also add something that basically hints that the setup is working fine. This runs the script, but we still wouldn't know if the integration works fine. We can add a script after running the script that ensures the setup works.

File: linux/amzn/2.0/foxpass_setup.py [hunk 5]
```diff
@@ -147,30 +169,109 @@ def configure_sssd(bind_dn, bind_pw, backup_ldaps, opt_timeout):
...
+    if debug:
+        to_file = open_file('/etc/sssd/sssd.conf')
+        diff_files(from_file, to_file, '/etc/sssd/sssd.conf')
+
+
+def configure_ldap_sudoers(base_dn, sudo_timed, full_refresh_interval, smart_refresh_interval, debug):
+    if debug:
+        from_file = open_file('/etc/sssd/sssd.conf')
+
+    from SSSDConfig import SSSDConfig
```
Comment (grk-it) [inline linux/amzn/2.0/foxpass_setup.py:181]: Can this go to the top of the file ?

File: linux/amzn/2.0/foxpass_setup.py [hunk 6]
```diff
@@ -147,30 +169,109 @@ def configure_sssd(bind_dn, bind_pw, backup_ldaps, opt_timeout):
...
+    sssdconfig = SSSDConfig()
+    sssdconfig.import_config('/etc/sssd/sssd.conf')
+
+    try:
+        sssdconfig.new_service('sudo')
+        sssdconfig.activate_service('sudo')
+    except:
+        pass
```
Comment (grk-it) [inline linux/amzn/2.0/foxpass_setup.py:189]: Under what scenarios we might encounter exceptions. Should we be handling them instead of letting it proceed ?

File: linux/debian/9/foxpass_setup.py [hunk 7]
```diff
@@ -61,12 +60,12 @@ def main():
 
     apt_get_update()
     install_dependencies()
-    write_foxpass_ssh_keys_script(apis, args.api_key)
-    write_nslcd_conf(uris, args.base_dn, binddn, args.bind_pw, args.ldap_connections, args.idle_timelimit)
-    augment_sshd_config()
-    augment_pam()
-    fix_nsswitch()
-    fix_sudo(args.sudoers_group, args.require_sudoers_pw, args.update_sudoers)
+    write_foxpass_ssh_keys_script(apis, args.api_key, args.debug)
```
Comment (grk-it) [inline linux/debian/9/foxpass_setup.py:63]: Instead of passing `debug` flag to all the functions, do you think it makes sense to check for the flag before calling the function and after calling the function for any differences to log. The debug in that case can just be handled here and needn't be in the individual function.

```
if debug:
    old = record_old_file('')
    write_foxpass_ssh_keys_script(apis, args.api_key)
    new = record_new_file('')
    record_diff(old, new)
```

File: linux/amzn/2.0/foxpass_setup.py (hunk 5 above)
Comment (asakapab0i) [reply linux/amzn/2.0/foxpass_setup.py:181]: Resolved as discussed.

File: linux/amzn/2.0/foxpass_setup.py (hunk 6 above)
Comment (asakapab0i) [reply linux/amzn/2.0/foxpass_setup.py:189]: Resolved as discussed.

File: docker/Docker_amazonlinux2 (hunk 4 above)
Comment (asakapab0i) [reply docker/Docker_amazonlinux2:16]: Resolved as discussed.

File: linux/ubuntu/20.04/foxpass_setup.py [hunk 8]
```diff
@@ -46,17 +47,41 @@ def main():
...
+    parser.add_argument('--query-timelimit', default=30, help='The amount of time, in seconds, to wait while performing an LDAP query.')
 
     args = parser.parse_args()
 
     binddn = 'cn=%s,%s' % (args.bind_user, args.base_dn)
     apis = [args.api_url] + args.apis
     uris = [args.ldap_uri] + args.ldaps
 
+    if args.debug:
+        foxpass_ssh_keys_path = '/usr/local/sbin/foxpass_ssh_keys.sh'
```
Comment (grk-it) [inline linux/ubuntu/20.04/foxpass_setup.py:65]: Can we create a function that opens various files and returns the file object to the caller. That way we don't have to have the same code replicated across different places.

Comment (asakapab0i) [reply linux/ubuntu/20.04/foxpass_setup.py:65]: Can you provide the example on this?

Comment (grk-it) [reply linux/ubuntu/20.04/foxpass_setup.py:65]: i think don't worry, seeing the same code across files i thought we could create a helper function. But then we need to have this script installed as part of the setup. Single standalone script is probably better.

**Review by grk-it — APPROVED** [review]

**Comment by grk-it** [comment]
Thanks Bryan for making all the changes!
